lib/scope.js

```diff
@@ -385,11 +383,11 @@ AST_Symbol.DEFMETHOD("mark_enclosed", function(options) {
     var s = this.scope;
     while (s) {
         push_uniq(s.enclosed, def);
-        if (options.keep_fnames) {
-            s.functions.each(function(d) {
+        s.functions.each(function(d) {
+            if (keep_name(options.keep_fnames, d.name)) {
```

Comment:
It's expensive to do the for/each when the non-default `keep_fnames` is enabled. Keep the outer `if (options.keep_fnames)` condition.